### DISCUSSION TITLE
I don't understand how to use server/api routes with pinia store

### DISCUSSION THREAD

**Discussion by melokki** [nuxt/nuxt#16665] [Questions] [answered]
Hello everyone

I've recently started to develop a application using `ssr`, but I'm facing some issues with the `server/api` routes and pinia store.

I've created in `server/api` folder a new file `categories.ts` where I have the following code 

```ts
export default eventHandler(async (event) => {
  return {
    data: [
      {
        name: "category 1",
      },
      { 
        name: "category 2" 
      },
    ],
  };
});
```

I've created the `store` folder with a category store in which I have the following store

```ts
export const useCategory = defineStore("category", {
  state: (): CategoriesState => {
    return {
      categories: [],
    };
  },
  actions: {
    async getAll(): Promise<void> {
      const { data } = await $fetch<{
        data: Category[];
      }>(`/api/categories/`);
      this.categories = data;
    },
  },
});
```

in my component I am trying to access the store action `getAll`  like this:

```ts
<script setup lang="ts">
onBeforeMount(async () => {
 await categories.getAll();
});
</script>
```

In my terminal I am getting the following warning

```
[Vue Router warn]: No match found for location with path "/api/categories/"
```

I don't understand how can I use the `/server/api` routes on the same time with the pinia store

If I call the `/api/categories` from the component like this:

```ts
<script setup lang="ts">
    const { data } = await useAsyncData('categories', () => $fetch('/api/categories'))
</script>
````

I can see the result in the console, but that's not like how I imagine the process, I would like to make the call from the pinia store actions

**Comment by TheAlexLichter** [comment, marked as answer]
In your call in Pinia, you use a trailing slash (`      }>('/api/categories/');`) which you should not.

Please see [my stackblitz](https://stackblitz.com/edit/https-github-comnuxtframeworkdiscussions8019) for a working example of your code and the [official example](https://stackblitz.com/github/piniajs/example-nuxt-3?file=server%2Fapi%2Fhello.ts).

**Reply by melokki** [reply]
Thank you very much for your reply, now it works partially as expected. Now it finds the route and it hits it, returns the correct result, but the route is called every time I navigate in the website. 
I placed some ifs to check the state so I can stop the call, but it's like I am hitting the refresh button, the state is always empty and the request is made

**Reply by TheAlexLichter** [reply]
Yes, this is normal because the state is not persisted on client-side. You can persist your store e.g. via localstorage *or* do the call on the server-side (not on `onBeforeMounted` which is only called on client-side) so it will be part of the initial response ☺️

**Reply by melokki** [reply]
Thank you very much for your time and for your reply, it's starting to be a little clearer

I decided to choose the second variant, to make the call on the server side. I've created the plugin directory, I've created a new plug in `getCategories.server.ts` and I used this code (I am not sure this is the right way of doing it)

```ts
import { useCategory } from "~~/store/categories";

export default defineNuxtPlugin(async (nuxtApp) => {
  const categories = useCategory(nuxtApp.$pinia);


  await categories.getAll();
});
```
and in my default layout I am accessing the the store property to list all the categories
it works, but it has the same problem, every time I navigate a link, it calls the server endpoint again. 
Here is the part I am not sure I understand, how can I make it to hit the server endpoint only once

**Reply by TheAlexLichter** [reply]
Then you must call `categories.getAll();` still somewhere else 🤔

**Reply by melokki** [reply]
understood, but where and how exactly :thinking:

**Reply by melokki** [reply]
Thank you very much again @manniL for your time and for your replays.
I am so stupid because I was navigating the website using the `<a>` element instead of using the `<NuxtLink>` component, I've change it now and works like a charm :sweat_smile:   thank you very much again  :pray:
